### ISSUE TITLE
Handling all events

### ISSUE THREAD

**Issue by impredicative** [luk3yx/miniirc#17]
I'd like to be able to have a handler that can handle all events rather than specific types of events. How would this be doable? I may use this to put all events on a message queue. Thanks.

Within such a handler, I would definitely also need to know as an input argument which kind of event is being handled. I don't believe this is currently passed as a function argument.

**Comment by luk3yx** [comment]
Can you clarify further how this would be useful? I've considered implementing it earlier, however I don't see much use for it.

**Comment by impredicative** [comment]
It would allow me to to just put the messages on a queue. The messages will then be selectively read by other processes, some of which may not even use Python. This is a low priority request, so we can just revisit it later if it becomes a high priority.

**Comment by luk3yx** [comment]
I see no reason why this should be closed, do you have any suggestions on how this should be implemented? The easiest way would probably be to create a special `*` handler:

```py
@irc.Handler('*')
def handler(irc, hostmask, args):
    print(hostmask[0], 'sent a', args[0], 'with contents', args[1:])

# >>> :nick!user@host PRIVMSG #channel :Test message
# nick sent a PRIVMSG with contents ['#channel', ':Test message']
```

Alternatively, a separate argument (`cmd`) could be sent to the handler, however that would add complexity.

**Comment by impredicative** [comment]
Even in the simpler case of handling multiple (but not all) events (e.g. both PING and PONG) via a single handler, I need to know which command the handler was called for. The solution for providing the command should therefore apply to all cases.

I'm okay with either one of the two solutions you suggested. If I am not mistaken, they will both break existing code, right?

**Comment by luk3yx** [comment]
`irc.Handler()` seems much nicer than `irc.Handler('*')`.

I could possibly add another keyword argument to `irc.Handler` that adds an extra `cmd` argument (and this would be on by default for handlers that handle all events).

**Comment by impredicative** [comment]
OK. I may order them as `(irc, cmd, hostmask, args)`, but I don't have a strong opinion on this.

**Comment by impredicative** [comment]
I definitely see the potential here for misuse, as per your example.

**Comment by luk3yx** [comment]
I am not entirely sure what to name the new keyword argument, at the moment I'm thinking about `cmd_arg`:

```py
@irc.Handler('PRIVMSG', 'NOTICE', cmd_arg = True)
def handler(irc, cmd, hostmask, args):
    pass
```

**Comment by luk3yx** [comment]
Created the (WIP) [`catch-all`](https://github.com/luk3yx/miniirc/tree/feature/catch-all) branch.

**Comment by impredicative** [comment]
I would just create a breaking 2.0 version and always send command to all handlers.

**Comment by luk3yx** [comment]
I'd rather not break backwards compatibility (at the moment) if possible, I'd have to update lots of code.

**Comment by impredicative** [comment]
It's better in the long term to have a clean API. Having to provide `cmd_arg=True`, and will only extend the pain forever.

**Comment by impredicative** [comment]
Here is a possible alternative, although I'm not entirely sure if it's a sound one:

```
@irc.BaseHandler('PRIVMSG', 'NOTICE')
def handler(irc, cmd, hostmask, args):
    pass
```

`BaseHandler` receives the command, always. The regular `Handler` doesn't. Both are callable as `()`.

**Comment by luk3yx** [comment]
I like the idea, however I'm unsure about the name `BaseHandler`.

**Comment by impredicative** [comment]
You're right. The prefix `Base` is meant to be suggestive of a class, so it doesn't quite fit here.

**Comment by luk3yx** [comment]
Currently (in the [`catch-all`](https://github.com/luk3yx/miniirc/tree/feature/catch-all) branch), calling `Handler()` without parameters is the same as calling `BaseHandler()` without parameters, and will call the function with the `cmd` parameter.

The only other names for `irc.BaseHandler` I can think of are `irc.on` and `irc.Event`, however they would probably get mixed up with `irc.Handler` easily.

**Comment by impredicative** [comment]
`irc.EventHandler` could be a possible alternative name.

I wouldn't use the arg name `cmd` because there is a built-in importable module with that name. I would use `command` or `event`.

**Comment by luk3yx** [comment]
`irc.CmdHandler` makes slightly more sense than `irc.BaseHandler`.

The name `command` is now used when calling handlers (and will be the name used in the documentation), however miniirc still uses `cmd` internally.

**Comment by impredicative** [comment]
`irc.CmdHandler` sounds fine to me.

To recap, it will receive the positional parameters: irc, command, hostmask, args

**Comment by luk3yx** [comment]
Does [the documentation](https://github.com/luk3yx/miniirc/tree/feature/catch-all#handling-multiple-events) explain this feature correctly?

**Comment by impredicative** [comment]
> Calling irc.Handler() without parameters is the same as calling irc.CmdHandler() without parameters.

But the `command` parameter is provided only to the latter, right? I'm just asking, not suggesting anything.

**Comment by luk3yx** [comment]
It's sent to both, I don't see much use for catch-all handlers without the `command` parameter.

**Comment by impredicative** [comment]
That's true.

**Comment by impredicative** [comment]
The problem there is that a tool like `mypy` will then have a very hard time analyzing this in the miniirc code.

**Comment by luk3yx** [comment]
Can you clarify how it's a problem?

It could possibly raise an exception if `irc.Handler()` is called without parameters.

**Comment by impredicative** [comment]
As we know, mypy is just a static analyzer, and it doesn't do any runtime checks. It's the variable type definition of the decorated callable that concerns me. It may be okay; I'm not totally sure. Fwiw, you can see the example of a type annotation for a `Callable` [here](https://mypy.readthedocs.io/en/latest/cheat_sheet_py3.html) and [here](https://docs.python.org/3/library/typing.html#typing.Callable).

**Comment by luk3yx** [comment]
Calling `irc.Handler()` without parameters will now raise a `TypeError`, it would probably get confusing if it passed a `command` argument anyway.

**Comment by luk3yx** [comment]
Released miniirc 1.3.0 with catch-all handlers and `CmdHandler`s.
